system/security/shared/caching.cpp

```diff
@@ -223,12 +227,27 @@ CPermissionsCache::~CPermissionsCache()
     flush();
 }
 
-int CPermissionsCache::lookup( ISecUser& sec_user, IArrayOf<ISecResource>& resources, 
-                            bool* pFound)
+int CPermissionsCache::lookup( ISecUser& sec_user, IArrayOf<ISecResource>& resources, bool* pFound)
 {
+    time_t tstamp;
+    time(&tstamp);
+
     const char* userId = sec_user.getName();
+
+    //First, clear stale cache entries for this CResPermissionsCache entry
+    {
+        WriteLockBlock writeLock(m_resPermCacheRWLock);
```

Comment:
New logic minimizes serialization down to lines 239-246, is there a a way to minimize it even further?
We discussed only requesting writelocks for the case where the user entry has expired. Let us know why that approach isn't better than the current approach.